### PR TITLE
Use CUDF_JNI_ENABLE_PROFILING to conditionally enable profiling support.

### PR DESCRIPTION
**Pull request by bdice**
## Description
The conda cudatoolkit packages do not include `<cuda_profiler_api.h>`. This PR makes CUDA profiling (introduced in #9543) optional in the JNI layer. By default, profiling will be enabled. In GitHub Actions (#12002), profiling will be disabled.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: java/src/main/native/CMakeLists.txt [hunk 1]
```diff
@@ -87,6 +88,10 @@ if(CUDF_USE_PER_THREAD_DEFAULT_STREAM)
   add_compile_definitions(CUDA_API_PER_THREAD_DEFAULT_STREAM CUDF_USE_PER_THREAD_DEFAULT_STREAM)
 endif()
 
+if(CUDF_JNI_ENABLE_PROFILING)
+  target_compile_definitions(cudfjni PUBLIC CUDF_JNI_ENABLE_PROFILING)
```
Comment (bdice) [inline java/src/main/native/CMakeLists.txt:92]: @rapidsai/rapids-cmake-codeowners Can someone look at this CMake? Do I need a value here or something like `"CUDF_JNI_ENABLE_PROFILING=1"`? I don't feel super confident in this.

**Comment by abellina** [comment]
The JNI side of this looks great, I am not too sure on the CMake side of things either.

File: java/src/main/native/CMakeLists.txt (hunk 1 above)
Comment (jbrennan333) [reply java/src/main/native/CMakeLists.txt:92]: I'm not a CMake expert, but this looks correct to me.

**Review by jbrennan333 — APPROVED** [review]
+1 this looks good to me

**Comment by abellina** [comment]
JNI build failed on:

> 08:13:45      [exec] CMake Error at CMakeLists.txt:92 (target_compile_definitions):
08:13:45      [exec]   Cannot specify compile definitions for target "cudfjni" which is not built
08:13:45      [exec]   by this project.

File: java/src/main/native/CMakeLists.txt (hunk 1 above)
Comment (jbrennan333) [reply java/src/main/native/CMakeLists.txt:92]: Based on the build failure, I think this should be `add_compile_definitions` like the code just above.

Comment (bdice) [reply java/src/main/native/CMakeLists.txt:92]: I am going to push a fix for this shortly. I think this CMake file has some issues. We _should_ prefer `target_compile_definitions` to keep the scope limited, but `add_library(cudfjni)` occurs lower in the script. All `target_compile_definitions` must happen after the library is defined. This is also an issue for `USE_NVTX`, which I was using as an example. I am guessing that we have not recently tested the build with that option turned off.

Comment (robertmaynard) [reply java/src/main/native/CMakeLists.txt:92]: both `CUDF_JNI_ENABLE_PROFILING` and `CUDF_JNI_ENABLE_PROFILING=1` are valid. It depends on if you want to define a name or a name with value.

```cpp
#define CUDF_JNI_ENABLE_PROFILING 
// versus
#define CUDF_JNI_ENABLE_PROFILING 1
```

Comment (robertmaynard) [reply java/src/main/native/CMakeLists.txt:92]: But yes this needs to go after the construction of the target `cudfjni`

**Review by robertmaynard — CHANGES_REQUESTED** [review]

File: java/src/main/native/CMakeLists.txt [hunk 2]
```diff
@@ -152,6 +148,15 @@ add_library(
   src/check_nvcomp_output_sizes.cu
 )
 
+# Disable NVTX if necessary
+if(NOT USE_NVTX)
+  target_compile_definitions(cudfjni PUBLIC NVTX_DISABLE)
+endif()
+
+if(CUDF_JNI_ENABLE_PROFILING)
+  target_compile_definitions(cudfjni PUBLIC CUDF_JNI_ENABLE_PROFILING)
```
Comment (bdice) [inline java/src/main/native/CMakeLists.txt:157]:
```suggestion
  target_compile_definitions(cudfjni PRIVATE CUDF_JNI_ENABLE_PROFILING)
```

File: java/src/main/native/CMakeLists.txt [hunk 3]
```diff
@@ -152,6 +148,15 @@ add_library(
   src/check_nvcomp_output_sizes.cu
 )
 
+# Disable NVTX if necessary
```
[the remark this answers is not in the document]
Comment (bdice) [reply java/src/main/native/CMakeLists.txt:151]: Discussed offline. `CUDF_JNI_ENABLE_PROFILING` should be private. `NVTX_DISABLE` probably needs to stay public (it's also public in the cudf target).

**Review by robertmaynard — APPROVED** [review]

**Comment by bdice** [comment]
@gpucibot merge

**Pull request merged**